src/ParallelAlgorithms/SurfaceFinder/SurfaceFinder.cpp

```diff
@@ -0,0 +1,109 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include <array>
+#include <cstddef>
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "ParallelAlgorithms/SurfaceFinder/SurfaceFinder.hpp"
+
+#include "DataStructures/ApplyMatrices.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"
+#include "Domain/Domain.hpp"
+#include "NumericalAlgorithms/RootFinding/TOMS748.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+
+// Wrapping for the interpolator for Toms748 rootfind.
+struct RayHelper {
+ public:
+  explicit RayHelper(DataVector ray)
+      : N_points_(ray.size()),
+        values_(ray),
+        extent_(Index<1>{std::array<size_t, 1>{N_points_}}) {}
+  double operator()(double x) {
+    Matrix int_mat =
+        Spectral::interpolation_matrix<Spectral::Basis::Legendre,
+                                       Spectral::Quadrature::GaussLobatto>(
+            N_points_, DataVector{1, x});
+    std::array<Matrix, 1> int_arr{int_mat};
+    DataVector result = apply_matrices(int_arr, values_, extent_);
+    return result[0];
+  }
+
+ private:
+  size_t N_points_;
+  DataVector values_;
+  Index<1> extent_;
+};
+
+namespace SurfaceFinder {
+
+std::vector<std::optional<double>> FindInElement(
+    const double target, Index<3>& extents,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& inertial_coords,
+    const Scalar<DataVector>& data, const DataVector& xhi_points,
+    const DataVector& eta_points) {
+  // Build the matrices for interpolation, as well as the result DataVector
+  const size_t xhi_number = xhi_points.size();
+  const size_t eta_number = eta_points.size();
+  const Matrix xhi_interp_matrix =
+      Spectral::interpolation_matrix<Spectral::Basis::Legendre,
+                                     Spectral::Quadrature::GaussLobatto>(
+          extents[0], xhi_points);
+  const Matrix eta_interp_matrix =
+      Spectral::interpolation_matrix<Spectral::Basis::Legendre,
+                                     Spectral::Quadrature::GaussLobatto>(
+          extents[1], eta_points);
+  std::array<Matrix, 3> interp_array;
+  interp_array[0] = xhi_interp_matrix;
+  interp_array[1] = eta_interp_matrix;
+  std::vector<std::optional<double>> result(xhi_number * eta_number);
+
+  // Interpolate the grid onto the rays, same with the radii. Subtract off the
+  // target contour.
+  const Scalar<DataVector> radii = magnitude(inertial_coords);
+  const Scalar<DataVector> interpolated_data{
+      apply_matrices(interp_array, data[0], extents)};
+  const Scalar<DataVector> interpolated_radii{
+      apply_matrices(interp_array, radii[0], extents)};
```

Comment:
Why do you need to interpolate the radii at all? I think from reading your code you're using it only to get the radius in inertial coords from the radius in logical coords. Instead, just return the radius in logical coordinates here. It can be mapped to inertial coords outside of this function using the coordinate maps. Then I think you also don't need the `inertial_coordinates` argument to this function anymore.